standard-libraries/list.3.mr

```diff
@@ -0,0 +1,69 @@
+## Generic lists
+
+## Generic single link list
+static List{Generic}
+  owner ListNode{Generic} first
+  user ListNode{Generic} last
+  
+  ## Add item to list end
+  func inst append(owner Generic item)
+    new ListNode{Generic} node
+    node.item := item
+    if ? self.last
+      self.last.next := node
+    else
+      self.first := node
+    self.last := node
+  
+  ## Add item to list start
+  func inst prepend(owner Generic item)
+    new ListNode{Generic} node
+    node.item := item
+    node.next := self.first
+    self.first := node
+    if not (? self.last)
+      self.last := node
+  
+  func inst is-empty():(copy Bool empty)
+    empty := not (? self.first)
+  
+  ## Get and remove first item, fail if list is empty
+  func inst pop():(owner Generic item)
+    if not (? self.first)
+      raise
+    owner ListNode{Generic} first(self.first)
+    item := first.item
+    self.first := first.next
+    delete first
+    if not (? self.first)
+      self.last := _
+
+
+## Iterate over the list
+class ListIterator{Generic}(Iterator)
+  user ListNode{Generic} curr
+  
+  ## Start iterating over given list
+  func inst init(user List{Generic} list)
+    self.curr := list.first
+  
+  ## Return whether current item exist
+  func dynamic has-item():(copy Bool exist)
+    exist := ? self.curr
+  
+  ## Get current item, fail if not exist
+  func dynamic get-item():(user Generic item)
+    if not (? self.curr)
```

Comment:
As we discussed, this would be far more readable as `if self.curr is not Empty` (you can also use `None` for more Python "compatability").